docs/integrations/custom_middleware.md

```diff
@@ -0,0 +1,234 @@
+# Writing your own Custom Middleware
+
+## How to write your own custom middleware for Mirascope
+
+`middleware_decorator` is a helper function to assist in helping you wrap any Mirascope call.
+We will be creating an example decorator `with_saving` that saves some metadata after a Mirascope call using [SQLModel](https://sqlmodel.tiangolo.com/). We will be using this table for demonstrative purposes in our example:
+
+```python
+from decimal import Decimal
+
+from sqlmodel import Field, SQLModel
+from sqlalchemy import JSON, Column
+
+class CallResponseTable(SQLModel, table=True):
+    """CallResponse model"""
+
+    __tablename__ = "call_response"
+    id: int | None = Field(default=None, primary_key=True)
+    function_name: str = Field(default="")
+    prompt_template: str | None = Field(default=None)
+    content: str | None = Field(default=None)
+    response_model: dict | None = Field(sa_column=Column(JSON), default=None)
+    cost: Decimal | None = Field(default=None)
+```
+
+This table should be adjusted and tailored to your needs depending on your SQL Dialect or requirements.
+
+### Writing the decorator
+
+```python
+from mirascope.integrations import middleware_decorator
+
+def with_saving(fn):
+    """Saves some data after a Mirascope call."""
+
+    return middleware_decorator(
+        fn,
+        custom_context_manager=custom_context_manager,
+        custom_decorator=custom_decorator,
+        handle_call_response=handle_call_response,
+        handle_call_response_async=handle_call_response_async,
+        handle_stream=handle_stream,
+        handle_stream_async=handle_stream_async,
+        handle_response_model=handle_response_model,
+        handle_response_model_async=handle_response_model_async,
+        handle_structured_stream=handle_structured_stream,
+        handle_structured_stream_async=handle_structured_stream_async,
+    )
+```
+
+We will go over each of the different functions.
+
+### `custom_context_manager`
+
+We start off with the `custom_context_manager` function which will be relevant to all the handlers. You can define your own context manager where the yielded value is passed to each of the handlers.
+
+```python
+from contextlib import contextmanager
+from typing import Any, Callable, Generator
+
+from sqlmodel import Session
+
+@contextmanager
+def custom_context_manager(
+    fn: Callable,
+) -> Generator[Session, Any, None]:
+    engine = ...
+    print(f"Saving call: {fn.__name__}")
+    with Session(engine) as session:
+        yield session
+```
+
+All the handlers will be wrapped by this context manager. The `fn` argument will be your Mirascope call function.
+
+### `handle_call_response` and `handle_call_response_async`
+
+These functions will be called after making a mirascope call with the following signature:
```

Comment:
wording here is a little weird (sounds like the mirascope call needs this signature).

Maybe something like the following is more clear?

"These functions must have the following signature (where async should be async) and will be called after making a standard Mirascope call. Here is a sample implementation of the sync version:"